contracts/Staking.sol

```diff
@@ -0,0 +1,177 @@
+
+    /// @notice the constant that store decimals 18
+    /// @dev the constant that store decimals 18
+    uint256 public constant DECIMALS18 = 10**18;
+
+    /// @notice Store the annual percentage yield
+    /// @dev Store the annual percentage yield
+    uint256 public apy;
+
+    /// @notice Store the start date of stake
+    /// @dev Store the start date of stake
+    uint256 public startDate;
+
+    /// @notice Store the finish date of stake
+    /// @dev Store the finish date of stake
+    uint256 public endDate;
+
+    /// @notice Store the total supply of staked tokens
+    /// @dev Store the total supply of staked tokens
+    uint256 public totalSupply;
+
+    /// @notice Store the ERC20 staking token
+    /// @dev Store the ERC20 staking token
+    IERC20 public stakingToken;
+
+    /// @notice Store the ERC20 reward token
+    /// @dev Store the ERC20 reward token
+    IERC20 public rewardToken;
+
+    /// @notice Store the info about staking
+    /// @dev Store the info about staking
+    StakingInfo public stakingInfo;
+
+    /// @notice Store the users staked amount tokens
+    /// @dev Store the users staked amount tokens
+    mapping(address => uint256) public staked;
+
+    /// @notice Store the users last staked time
+    /// @dev Store the users last staked time
+    mapping(address => uint256) public stakeTime;
+
+    /// @notice Store the users reward tokens
+    /// @dev Store the users reward tokens
+    mapping(address => uint256) public rewards;
+
+    /// @notice Initialize contract
+    /// @dev Initialize contract, sets reward/staking token addresses & staking info
+    /// @param rewardToken_ the reward token address
+    /// @param stakingToken_ the staking token address
+    constructor(address rewardToken_, address stakingToken_) {
+        require(
+            rewardToken_ != address(0) && stakingToken_ != address(0),
+            "ERROR_INVALID_ADDRESS"
+        );
+        rewardToken = IERC20(rewardToken_);
+        stakingToken = IERC20(stakingToken_);
+        stakingInfo.feePercentage = 40;
+        stakingInfo.cooldown = 10 days;
+        stakingInfo.stakingPeriod = 60 days;
+    }
+
+    /// @notice Add reward to contract for a specific period
+    /// @dev  Add reward to contract for a specific period
+    /// @param _start the start date of stake
+    /// @param _finish the finish date of stake
+    /// @param _rewardsAmount the amount of reward tokens
+    /// @param _apy the annual percentage yield
+    function setRewards(
+        uint256 _start,
+        uint256 _finish,
+        uint256 _rewardsAmount,
+        uint256 _apy
+    ) external virtual override onlyOwner {
+        require(
+            _start < _finish && block.timestamp < _start,
+            "Incorrect dates"
+        );
+        require(
+            _rewardsAmount != 0 && _rewardsAmount <= 500_000 * DECIMALS18,
+            "Incorrect amount of tokens"
+        );
+        require(_apy != 0 && _apy <= 10, "Incorrect apy");
+
```

Comment:
you can set _apy 10 % and set rewards 200 000 . then when all user stake ( 5 000 000 tokens) and have apy 10%. They will have 500 000 rewards but we add only 200 000 rewards tokens. we need avoid this case